<a href="https://colab.research.google.com/github/shinkarkunal/EDA_Hotel_booking/blob/main/Hotel_Booking_Analysis_Capstone_Project_KUNAL_SHINKAR_ipynb.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


Have you ever wondered when the best time of year to book a hotel room is? Or the optimal length of stay in order to get the best daily rate? What if you wanted to predict whether or not a hotel was likely to receive a disproportionately high number of special requests? This hotel booking dataset can help you explore those questions!
This data set contains booking information for a city hotel and a resort hotel, and includes information such as when the booking was made, length of stay, the number of adults, children, and/or babies, and the number of available parking spaces, among other things. All personally identifying information has been removed from the data.
Explore and analyze the data to discover important factors that govern the bookings

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


### Import required libraries

In [2]:
# importing necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
colors = ['#4F6272', '#B7C3F3', '#DD7596', '#8EB897']

### Load Dataset

In [3]:
# loading dataframe
hotel_df = pd.read_csv('/content/drive/MyDrive/Almabetter/EDA_Project/Hotel Bookings.csv')

In [4]:
hotel_df.head()

hotel  is_canceled  ...  reservation_status  reservation_status_date
0  Resort Hotel            0  ...           Check-Out               2015-07-01
1  Resort Hotel            0  ...           Check-Out               2015-07-01
2  Resort Hotel            0  ...           Check-Out               2015-07-02
3  Resort Hotel            0  ...           Check-Out               2015-07-02
4  Resort Hotel            0  ...           Check-Out               2015-07-03

[5 rows x 32 columns]

In [5]:
hotel_df.shape

(119390, 32)

In [6]:
hotel_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

In [7]:
hotel_df.describe()

is_canceled  ...  total_of_special_requests
count  119390.000000  ...              119390.000000
mean        0.370416  ...                   0.571363
std         0.482918  ...                   0.792798
min         0.000000  ...                   0.000000
25%         0.000000  ...                   0.000000
50%         0.000000  ...                   0.000000
75%         1.000000  ...                   1.000000
max         1.000000  ...                   5.000000

[8 rows x 20 columns]

### Data Cleaning

##### Before exploratory analysis, the data has to be cleaned first. Cleaning of Data basically deals with Null values and Outliers.In this we try to figure out those columns which is not adding any value in our analysis and remove such columns if any. Sometimes the data might contain information that does not make any sense. These have to be fixed too.

### Handling the Null values

##### In some columns values are missing (NAN value) But before move forword for analysis we have to fill those null values or delete column which contains high amount of null values for getting better insights from Data.

In [8]:
# sum of null values in each column
hotel_df.isna().sum().sort_values(ascending=False)

company                           112593
agent                              16340
country                              488
children                               4
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
hotel                                  0
previous_cancellations                 0
days_in_waiting_list                   0
customer_type                          0
adr                                    0
required_car_parking_spaces            0
total_of_special_requests              0
reservation_status                     0
previous_bookings_not_canceled         0
is_repeated_guest                      0
is_canceled                            0
distribution_channel                   0
market_segment                         0
meal                                   0
babies                                 0
adults                                 0
stays_in_week_ni

In [9]:
# lets try to find out percentage of null values present in columns
def percent_null_values(hotel_df):
  nan_percent = (hotel_df.isnull().sum()/len(hotel_df))*100
  nan_percent = nan_percent[nan_percent>0].sort_values(ascending=False)
  return nan_percent
percent_null_values(hotel_df)


company     94.306893
agent       13.686238
country      0.408744
children     0.003350
dtype: float64

##### Company Column has 94 % of Null values so its not feasible to fill that much null values so its better to drop this column.

In [10]:
# Drop the company column
hotel_df.drop(['company'], axis = 1, inplace = True)

##### Now lets deal with Agent Column

In [11]:
#just check From each agent how much hotels were booked
hotel_df["agent"].value_counts()

9.0      31961
240.0    13922
1.0       7191
14.0      3640
7.0       3539
         ...  
289.0        1
432.0        1
265.0        1
93.0         1
304.0        1
Name: agent, Length: 333, dtype: int64

##### From above data we have see that Maximum number of hotels book by agent number or agent_id 9 so that we will all null values with 9

In [12]:
# Replace NAN value with 9
hotel_df['agent'] = hotel_df['agent'].fillna(9)

In [13]:
# again check From each agent how much hotels were booked
hotel_df["agent"].value_counts()

9.0      48301
240.0    13922
1.0       7191
14.0      3640
7.0       3539
         ...  
280.0        1
285.0        1
289.0        1
265.0        1
497.0        1
Name: agent, Length: 333, dtype: int64

##### From above information we can see that agent with ID 9 count are increased 39961 to 48301

In [14]:
# dropping the rows with null values completely
hotel_df = hotel_df.dropna(axis=0)

##### For dealing with children and counntry column we drop those rows which has null values

In [15]:
# sum of null values in each column
hotel_df.isna().sum().sort_values(ascending=False)

hotel                             0
is_repeated_guest                 0
reservation_status                0
total_of_special_requests         0
required_car_parking_spaces       0
adr                               0
customer_type                     0
days_in_waiting_list              0
agent                             0
deposit_type                      0
booking_changes                   0
assigned_room_type                0
reserved_room_type                0
previous_bookings_not_canceled    0
previous_cancellations            0
distribution_channel              0
is_canceled                       0
market_segment                    0
country                           0
meal                              0
babies                            0
children                          0
adults                            0
stays_in_week_nights              0
stays_in_weekend_nights           0
arrival_date_day_of_month         0
arrival_date_week_number          0
arrival_date_month          

In [16]:
hotel_df.head()

hotel  is_canceled  ...  reservation_status  reservation_status_date
0  Resort Hotel            0  ...           Check-Out               2015-07-01
1  Resort Hotel            0  ...           Check-Out               2015-07-01
2  Resort Hotel            0  ...           Check-Out               2015-07-02
3  Resort Hotel            0  ...           Check-Out               2015-07-02
4  Resort Hotel            0  ...           Check-Out               2015-07-03

[5 rows x 31 columns]

##### From the describe() function used earlier we spotted something fishy. The minimum value of adults column is zero. While it is possible that there are bookings with children only it is better to check those data carefully.Dataset might contain bookings with zero adults,childrens.

In [17]:
# Check Bookings with zero adults
zero_adults = hotel_df[hotel_df.adults == 0]

In [18]:
zero_adults.head()

hotel  is_canceled  ...  reservation_status  reservation_status_date
2224  Resort Hotel            0  ...           Check-Out               2015-10-06
2409  Resort Hotel            0  ...           Check-Out               2015-10-12
3181  Resort Hotel            0  ...           Check-Out               2015-11-23
3684  Resort Hotel            0  ...           Check-Out               2016-01-04
3708  Resort Hotel            0  ...           Check-Out               2016-01-05

[5 rows x 31 columns]

In [19]:
# checking the shape of zero_adults
zero_adults.shape

(393, 31)

##### From above discussion We come to know that atleast a few of those data with zero adults have both children and babies zero as well. Which would imply that the total number of guests are zero and That does not make any sense. These data has to be considered as incorrect and will have to be dropped.It will be better to make a new column where these three categories are added together as 'all_guests' as compared to remove from data.

In [20]:
# create new column with all_guest
hotel_df['all_guests'] = hotel_df['adults'] + hotel_df['children'] + hotel_df['babies']

In [21]:
# check value counts for all_guest
hotel_df['all_guests'].value_counts()

2.0     81880
1.0     22289
3.0     10479
4.0      3926
0.0       170
5.0       137
26.0        5
12.0        2
27.0        2
20.0        2
10.0        2
40.0        1
50.0        1
55.0        1
6.0         1
Name: all_guests, dtype: int64

##### From above information we can see that we have total 170 observation with zero guest and its not make any sense so that We will drop all the data with zero total guests

In [22]:
# drop observation with zero adults
hotel_df.drop(hotel_df[hotel_df['all_guests']==0].index , inplace =  True)

In [23]:
# viewing those rows with zero all_guests (to confirm)
hotel_df[hotel_df.all_guests == 0].head()

Empty DataFrame
Columns: [hotel, is_canceled, lead_time, arrival_date_year, arrival_date_month, arrival_date_week_number, arrival_date_day_of_month, stays_in_weekend_nights, stays_in_week_nights, adults, children, babies, meal, country, market_segment, distribution_channel, is_repeated_guest, previous_cancellations, previous_bookings_not_canceled, reserved_room_type, assigned_room_type, booking_changes, deposit_type, agent, days_in_waiting_list, customer_type, adr, required_car_parking_spaces, total_of_special_requests, reservation_status, reservation_status_date, all_guests]
Index: []

##### Now we have data which does not have any null values or Irrelevent Data , we are ready to move forword